[SPARK-42617][PS] Support `isocalendar` from the pandas 2.0.0

dzhigimont opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Support `isocalendar` from the pandas 2.0.0

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
When pandas 2.0.0 is released, we should match the behavior in pandas API on Spark.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Added new method `DatetimeIndex.isocalendar` and removed two depreceted `DatetimeIndex.week` and `DatetimeIndex.weekofyear`
```
dfs = ps.from_pandas(pd.date_range(start='2019-12-29', freq='D', periods=4).to_series())
dfs.dt.isocalendar()
                    year  week  day
        2019-12-29  2019    52    7
        2019-12-30  2020     1    1
        2019-12-31  2020     1    2
        2020-01-01  2020     1    3
dfs.dt.isocalendar().week
        2019-12-29    52
        2019-12-30     1
        2019-12-31     1
        2020-01-01     1
```

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
UT was updated

Commit:
Implement isocalendar method

Commit:
Update documentation

Commit:
Update unit tests

Commit:
Remove redundant import

HyukjinKwon commented:
cc @itholic

HyukjinKwon commented on python/pyspark/pandas/datetimes.py:
```diff
@@ -116,26 +118,50 @@ def pandas_microsecond(s) -> ps.Series[np.int64]:  # type: ignore[no-untyped-def
+        Returns
+        -------
+        DataFrame
+            With columns year, week and day.
```
```suggestion
            With columns year, week and day.

```

HyukjinKwon replied:
Otherwise, the HTML rendering of this doc is broken.

HyukjinKwon commented on python/pyspark/pandas/datetimes.py:
```diff
@@ -116,26 +118,50 @@ def pandas_microsecond(s) -> ps.Series[np.int64]:  # type: ignore[no-untyped-def
-        warnings.warn(
-            "weekofyear and week have been deprecated.",
-            FutureWarning,
+        Calculate year, week, and day according to the ISO 8601 standard.
```
```suggestion
        Calculate year, week, and day according to the ISO 8601 standard.

```

Commit:
Fix doc string

Commit:
Fix doc string for the isocalendar

Co-authored-by: Hyukjin Kwon <gurwls223@gmail.com>

Commit:
Fix doc string for the isocalendar

Co-authored-by: Hyukjin Kwon <gurwls223@gmail.com>

Commit:
Remove unused import

HyukjinKwon commented:
Cc @Yikun too if you find some time to review.

dzhigimont committed:
add versioneadded for isocalendar

HyukjinKwon commented:
cc @itholic @zhengruifeng @xinrong-meng if you find some time to review.

itholic commented:
Could you resolve mypy check? You can run the static analysis by running `dev/lint-python` locally.

dzhigimont commented:
> Could you resolve mypy check? You can run the static analysis by running `dev/lint-python` locally.

@itholic  I've fixed the issue with mypy in my changes, but I still see three problems that don't relate to changes. Should I fix them?

```
annotations failed mypy checks:
python/pyspark/pandas/namespace.py:162: error: Cannot assign multiple types to name "_range" without an explicit "Type[...]" annotation  [misc]
python/pyspark/pandas/indexes/base.py:2093: error: unused "type: ignore" comment
python/pyspark/pandas/indexes/base.py:2163: error: unused "type: ignore" comment
Found 3 errors in 2 files (checked 507 source files)
```

itholic commented:
Can you rebase to master and try running linter again??
If the problem still exists, yes, let's fix the `mypy` failure. We should make the PR pass the CI anyway.

bjornjorgensen commented:
@dzhigimont we have upgraded main branch to pandas 2.0.2 now.

itholic commented:
@dzhigimont Could you proceed this PR if you're still interested on this work?

dzhigimont commented:
> @dzhigimont Could you proceed this PR if you're still interested on this work?

Sure, sorry for the delay I was a little bit overloaded on my work

itholic commented:
No worries! I ping you just for reminder :-) Please take your time

Commit:
Rebase to master and update tests

itholic commented on python/pyspark/pandas/datetimes.py:
```diff
@@ -116,26 +117,55 @@ def pandas_microsecond(s) -> ps.Series[np.int32]:  # type: ignore[no-untyped-def
+        2019-12-29  2019    52    7
+        2019-12-30  2020     1    1
+        2019-12-31  2020     1    2
+        2020-01-01  2020     1    3
```
nit: Could you add a new line between each examples to split them when displaying in the documents.

dzhigimont replied:
Added

itholic commented on python/pyspark/pandas/datetimes.py:
```diff
@@ -116,26 +117,55 @@ def pandas_microsecond(s) -> ps.Series[np.int32]:  # type: ignore[no-untyped-def
+        def pandas_isocalendar(  # type: ignore[no-untyped-def]
+            pdf,
+        ) -> ps.DataFrame[return_types]:  # type: ignore[valid-type]
+            # cast to int64 due to UInt32 is not supported by spark
```
> cast to int64 due to UInt32 is not supported by spark
 
Is this mean that the result is different from pandas ?? If so, let's add a "Note" to the docstring so that users recognize this difference instead of just adding the comment here.

dzhigimont replied:
Added a note

itholic commented on python/pyspark/pandas/tests/indexes/test_datetime.py:
```diff
@@ -101,22 +102,8 @@ def test_properties(self):
                 self.assert_eq(psidx.day_of_week, pidx.day_of_week)
 
         if LooseVersion(pd.__version__) >= LooseVersion("2.0.0"):
```
Let's remove this to focus on testing for the latest pandas version.

dzhigimont replied:
Removed

itholic commented on python/pyspark/pandas/tests/indexes/test_datetime.py:
```diff
@@ -269,6 +256,10 @@ def test_map(self):
         mapper_pser = pd.Series([1, 2, 3], index=pidx)
         self.assert_eq(psidx.map(mapper_pser), pidx.map(mapper_pser))
 
+    def test_isocalendar(self):
```
If we want to add a new test, then I think it's better to also move the related tests from "test_properties" to here.

dzhigimont replied:
Moved to the appropriate test

Commit:
Address feedback

Commit:
Remove unused import

Commit:
Fix the note format

Commit:
Fix datetime_property test

itholic commented on python/pyspark/pandas/datetimes.py:
```diff
@@ -116,26 +117,59 @@ def pandas_microsecond(s) -> ps.Series[np.int32]:  # type: ignore[no-untyped-def
+            is not supported by spark
+
+        Examples
+        --------
```
Can we consider & have an example including `pd.NaT`? Seems like this case is not working currently:

**Pandas**
```python
>>> ser = pd.to_datetime(pd.Series(["2010-01-01", pd.NaT]))
>>> ser.dt.isocalendar()
   year  week  day
0  2009    53     5
1  <NA>  <NA>  <NA>
```

**Current implementation**
```python
>>> ser = pd.to_datetime(pd.Series(["2010-01-01", pd.NaT]))
>>> psser = ps.from_pandas(ser)
# ValueError: cannot convert NA to integer
```

In Spark, we can't use mixed type within single column, so we should convert `NA` to proper type (e.g. use NaN instead of `NA` for float type in this case) as below:

```python
>>> psser.dt.week
0    53.0
1     NaN
dtype: float64
```

itholic replied:
@dzhigimont Could you take a look at this comment when you find some time? We can do it in separate PR in the future if you have not enough time for this for now :-)

dzhigimont replied:
If you give me 1-2 days I will try to resolve the proble

itholic replied:
Sure! It's not very urgent, so please take your time

itholic replied:
We can merge it before Spark 4.0.0, so we have enough time tho.

itholic commented on python/pyspark/pandas/indexes/datetimes.py:
```diff
@@ -214,28 +215,8 @@ def microsecond(self) -> Index:
-        return Index(self.to_series().dt.weekofyear)
-
-    weekofyear.__doc__ = week.__doc__
+    def isocalendar(self) -> DataFrame:
```
Do we need docstring?

dzhigimont replied:
Added

itholic reviewed (commented):
Otherwise looks pretty good

itholic commented:
@dzhigimont Can we just make the CI pass for now? I can help in the follow-ups after merging this one.

Seems like the mypy checks is failing for now:
```
starting mypy annotations test...
annotations failed mypy checks:
python/pyspark/pandas/namespace.py:162: error: Cannot assign multiple types to name "_range" without an explicit "Type[...]" annotation  [misc]
python/pyspark/pandas/indexes/base.py:2075: error: Unused "type: ignore" comment
python/pyspark/pandas/indexes/base.py:2145: error: Unused "type: ignore" comment
Found 3 errors in 2 files (checked 703 source files)
```

To resolve them:
- [ ] Remove "type: ignore" comment from python/pyspark/pandas/indexes/base.py:2075
- [ ] Remove "type: ignore" comment from python/pyspark/pandas/indexes/base.py:2145
- [ ] Add "# type: ignore" to python/pyspark/pandas/namespace.py:162

Commit:
Add doc for the isocalendar in DatetimeIndex

Commit:
Fix mypy checks

dzhigimont commented:
> @dzhigimont Can we just make the CI pass for now? I can help in the follow-ups after merging this one.
> 
> Seems like the mypy checks is failing for now:
> 
> ```
> starting mypy annotations test...
> annotations failed mypy checks:
> python/pyspark/pandas/namespace.py:162: error: Cannot assign multiple types to name "_range" without an explicit "Type[...]" annotation  [misc]
> python/pyspark/pandas/indexes/base.py:2075: error: Unused "type: ignore" comment
> python/pyspark/pandas/indexes/base.py:2145: error: Unused "type: ignore" comment
> Found 3 errors in 2 files (checked 703 source files)
> ```
> 
> To resolve them:
> 
> * [ ]  Remove "type: ignore" comment from python/pyspark/pandas/indexes/base.py:2075
> * [ ]  Remove "type: ignore" comment from python/pyspark/pandas/indexes/base.py:2145
> * [ ]  Add "# type: ignore" comment to python/pyspark/pandas/namespace.py:162

Fixed

dzhigimont commented:
> @dzhigimont Can we just make the CI pass for now? I can help in the follow-ups after merging this one.
> 
> Seems like the mypy checks is failing for now:
> 
> ```
> starting mypy annotations test...
> annotations failed mypy checks:
> python/pyspark/pandas/namespace.py:162: error: Cannot assign multiple types to name "_range" without an explicit "Type[...]" annotation  [misc]
> python/pyspark/pandas/indexes/base.py:2075: error: Unused "type: ignore" comment
> python/pyspark/pandas/indexes/base.py:2145: error: Unused "type: ignore" comment
> Found 3 errors in 2 files (checked 703 source files)
> ```
> 
> To resolve them:
> 
> * [ ]  Remove "type: ignore" comment from python/pyspark/pandas/indexes/base.py:2075
> * [ ]  Remove "type: ignore" comment from python/pyspark/pandas/indexes/base.py:2145
> * [ ]  Add "# type: ignore" comment to python/pyspark/pandas/namespace.py:162

I can't understand when I added Add "# type: ignore" comment to python/pyspark/pandas/namespace.py:162 mypy raised an error unused type: ignore when I deleted it raised   `python/pyspark/pandas/namespace.py:161: error: Cannot assign multiple types to name "_range" without an explicit "Type[...]" annotation  [misc]`
what do you suggest to me? How i can resolve the problem

itholic commented:
Yeah, mypy check always tricky 😂  Let me take a look

itholic commented on python/pyspark/pandas/datetimes.py:
```diff
@@ -116,26 +117,59 @@ def pandas_microsecond(s) -> ps.Series[np.int32]:  # type: ignore[no-untyped-def
+        Name: week, dtype: int64
+        """
+
+        return_types = [self._data.index.dtype, int, int, int]
+
+        def pandas_isocalendar(  # type: ignore[no-untyped-def]
+            pdf,
+        ) -> ps.DataFrame[return_types]:  # type: ignore[valid-type]
```
```suggestion
        def pandas_isocalendar(
            pdf: pd.DataFrame,
        ) -> ps.DataFrame[Any]: 
```

itholic reviewed (commented):
Could you also try the for `mypy` failure in `namespace.py`:
```diff
-_range = range  # built-in range
+# built-in range
+_range: Type[range] = range  # type: ignore[assignment]
```

itholic commented on python/pyspark/pandas/datetimes.py:
```diff
@@ -116,26 +117,57 @@ def pandas_microsecond(s) -> ps.Series[np.int32]:  # type: ignore[no-untyped-def
+        Name: week, dtype: int64
+        """
+
+        def pandas_isocalendar(
+            pdf: pd.DataFrame,
+        ) -> ps.DataFrame[Any]:
```
Oh, sorry we can't use such typing for Pandas UDF. Let's go back to the previous way.
```suggestion
        return_types = [self._data.index.dtype, int, int, int]

        def pandas_isocalendar(
            pdf: pd.DataFrame,
        ) -> ps.DataFrame[return_types]:  # type: ignore[valid-type]
```

itholic reviewed (approved):
Looks pretty good once CI pass.

itholic commented:
LGTM now.

Thanks for your consistent work on this!

HyukjinKwon commented:
Merged to master.
